File: web/src/shared/sim/killstreak-runtime.ts
/**
 * Killstreaks, made real.
 *
 * The definitions in data/killstreaks.ts describe what a streak costs and what
 * it claims to do; this is what actually happens when someone calls one in.
 *
 * The design rule throughout: **a killstreak must be counterable.** Anything
 * that puts damage on the map has a warning, a delay, and a way to be shot down
 * or avoided — a reward that simply deletes the enemy team stops being exciting
 * the second time you are on the receiving end. So the airstrike paints its
 * target and takes four seconds to arrive, the chopper can be shot out of the
 * sky, and the UAV gives away that it is up.
 */

import { anglesToForward, v3distance, v3set, vec3, type Vec3 } from '../math.js';
import type { Rng } from '../rng.js';
import {
  KillstreakVehicleKind,
  SimEventType,
  Team,
  isEnemyTeam,
  type EntityId,
  type KillstreakEntityState,
  type PlayerId,
  type PlayerState,
  type SimEvent,
  type WorldState,
} from '../types.js';
import { KILLSTREAKS, type KillstreakDef } from '../data/killstreaks.js';
import type { CollisionWorld } from '../collision/collision-types.js';
import { CollisionLayer, type QueryFilter } from '../collision/collision-types.js';

// ---------------------------------------------------------------------------
// Team-wide effects
// ---------------------------------------------------------------------------

/**
 * Effects that apply to a whole team rather than to an entity in the world.
 * Kept separate because they have no position, cannot be shot down, and every
 * consumer (minimap, HUD, AI) asks "is this active for my team" rather than
 * looking for an entity.
 */
export interface TeamEffects {
  /** Seconds of UAV coverage remaining — reveals enemies on the minimap. */
  uav: number;
  /** Advanced UAV also shows facing direction. */
  advancedUav: number;
  /** Counter-UAV jams the ENEMY minimap for this long. */
  counterUav: number;
  /** EMP disables enemy HUD and equipment. */
  emp: number;
}

function emptyEffects(): TeamEffects {
  return { uav: 0, advancedUav: 0, counterUav: 0, emp: 0 };
}

export interface KillstreakRuntime {
  effects: Map<Team, TeamEffects>;
  /** Scripted strikes in flight, resolved when their delay expires. */
  pendingStrikes: PendingStrike[];
  /** Next id for entities this system spawns. */
  nextId: number;
}

interface PendingStrike {
  kind: 'airstrike' | 'cluster' | 'cruise';
  owner: PlayerId;
  team: Team;
  target: Vec3;
  /** Seconds until impact. */
  delay: number;
  damage: number;
  radius: number;
  /** How many separate explosions this strike delivers. */
  bombs: number;
  /** Seconds between successive explosions in a run. */
  spacing: number;
  /** Which bomb we are up to. */
  fired: number;
  /** Direction the run travels, so bombs walk across the target. */
  heading: Vec3;
}

export function createKillstreakRuntime(): KillstreakRuntime {
  return {
    effects: new Map([
      [Team.Allies, emptyEffects()],
      [Team.Axis, emptyEffects()],
      [Team.None, emptyEffects()],
    ]),
    pendingStrikes: [],
    nextId: 50000,
  };
}

export function teamEffects(rt: KillstreakRuntime, team: Team): TeamEffects {
  let e = rt.effects.get(team);
  if (!e) {
    e = emptyEffects();
    rt.effects.set(team, e);
  }
  return e;
}

/**
 * Whether a team's minimap currently reveals enemies.
 *
 * Counter-UAV beats UAV: jamming is the direct counter to reconnaissance, and
 * making the newer or more expensive streak win instead would mean the counter
 * only works when you are already ahead.
 */
export function hasRadar(rt: KillstreakRuntime, team: Team): boolean {
  const own = teamEffects(rt, team);
  if (own.uav <= 0 && own.advancedUav <= 0) return false;
  for (const [other, effects] of rt.effects) {
    if (other === team || !isEnemyTeam(team, other)) continue;
    if (effects.counterUav > 0) return false;
  }
  return true;
}

// ---------------------------------------------------------------------------
// Calling one in
// ---------------------------------------------------------------------------

export interface KillstreakResult {
  events: SimEvent[];
  /** Entities to add to the world. */
  spawned: KillstreakEntityState[];
  /** True if the streak was consumed. */
  used: boolean;
  /** Set when the streak ends the match outright (the nuke). */
  endsMatch: boolean;
}

const _result: KillstreakResult = { events: [], spawned: [], used: false, endsMatch: false };
const _aim = vec3();
const _eye = vec3();

const STRIKE_FILTER: QueryFilter = { layers: CollisionLayer.World | CollisionLayer.Breakable };

/**
 * Call in a killstreak on behalf of a player.
 *
 * Offensive streaks are aimed: the target is wherever the player is looking,
 * traced onto the world. That is deliberate — it makes calling one a decision
 * rather than a button, and it means a badly-placed airstrike is the caller's
 * fault rather than the game's.
 */
export function callKillstreak(
  world: WorldState,
  collision: CollisionWorld,
  rt: KillstreakRuntime,
  player: PlayerState,
  streakId: string,
  rng: Rng,
): KillstreakResult {
  _result.events = [];
  _result.spawned = [];
  _result.used = false;
  _result.endsMatch = false;

  const def = KILLSTREAKS[streakId];
  if (!def) return _result;

  const idx = player.killstreakInventory.indexOf(streakId);
  if (idx < 0) return _result;

  // Where is the player pointing?
  v3set(_eye, player.position.x, player.position.y + 1.6, player.position.z);
  anglesToForward(_aim, player.yaw, player.pitch);
  const target = traceToGround(collision, _eye, _aim);

  switch (streakId) {
    case 'uav':
      teamEffects(rt, player.team).uav = def.duration;
      break;

    case 'advanced_uav':
      teamEffects(rt, player.team).advancedUav = def.duration;
      teamEffects(rt, player.team).uav = def.duration;
      break;

    case 'counter_uav':
      teamEffects(rt, player.team).counterUav = def.duration;
      break;

    case 'emp_burst':
      // EMP hits every enemy team.
      for (const [team, effects] of rt.effects) {
        if (isEnemyTeam(player.team, team)) effects.emp = def.duration;
      }
      break;

    case 'precision_airstrike':
      rt.pendingStrikes.push({
        kind: 'airstrike',
        owner: player.id,
        team: player.team,
        target,
        // Long enough that the warning is actionable. A strike that lands
        // before anyone can move is not a killstreak, it is a cutscene.
        delay: 4.0,
        damage: def.damage ?? 180,
        radius: def.radius ?? 7,
        bombs: 6,
        spacing: 0.18,
        fired: 0,
        heading: strikeHeading(player),
      });
      break;

    case 'cluster_strike':
      rt.pendingStrikes.push({
        kind: 'cluster',
        owner: player.id,
        team: player.team,
        target,
        delay: 3.0,
        damage: def.damage ?? 130,
        radius: def.radius ?? 5,
        bombs: 9,
        spacing: 0.12,
        fired: 0,
        heading: strikeHeading(player),
      });
      break;

    case 'cruise_missile':
      rt.pendingStrikes.push({
        kind: 'cruise',
        owner: player.id,
        team: player.team,
        target,
        delay: 2.5,
        damage: def.damage ?? 220,
        radius: def.radius ?? 9,
        bombs: 1,
        spacing: 0,
        fired: 0,
        heading: strikeHeading(player),
      });
      break;

    case 'sentry_gun':
    case 'care_package':
      // Deployables are placed at the player's feet by the caller; this system
      // only needs to record that the streak was spent.
      break;

    case 'attack_chopper':
    case 'chopper_gunner':
    case 'vtol_jet':
    case 'gunship': {
      const entity = spawnAirVehicle(rt, def, player, world);
      _result.spawned.push(entity);
      break;
    }

    case 'juggernaut':
      // Armour is applied to the caller directly and is the whole reward.
      player.maxHealth = 400;
      player.health = 400;
      break;

    case 'tactical_nuke':
      _result.endsMatch = true;
      break;

    default:
      break;
  }

  player.killstreakInventory.splice(idx, 1);
  _result.used = true;

  _result.events.push({
    type: SimEventType.KillstreakCalled,
    tick: world.tick,
    player: player.id,
    team: player.team,
    data: { killstreakId: streakId, target: { ...target } },
  } as SimEvent);

  // Both teams hear about it, but they hear different things.
  _result.events.push({
    type: SimEventType.Announce,
    tick: world.tick,
    team: player.team,
    line: def.friendlyAnnounce,
  } as SimEvent);
  if (def.enemyAnnounce) {
    _result.events.push({
      type: SimEventType.Announce,
      tick: world.tick,
      team: opposing(player.team),
      line: def.enemyAnnounce,
    } as SimEvent);
  }

  void rng;
  return _result;
}

function opposing(team: Team): Team {
  return team === Team.Allies ? Team.Axis : Team.Allies;
}

/** The direction a bombing run travels, so bombs walk away from the caller. */
function strikeHeading(player: PlayerState): Vec3 {
  const h = vec3();
  anglesToForward(h, player.yaw, 0);
  return h;
}

/** Trace the player's aim onto the world; falls back to a point ahead of them. */
function traceToGround(collision: CollisionWorld, origin: Vec3, dir: Vec3): Vec3 {
  const hit = collision.raycast(origin, dir, 300, STRIKE_FILTER, _traceHit);
  if (hit.hit) return vec3(hit.point.x, hit.point.y, hit.point.z);

  // Looking at the sky: drop the strike 60m ahead at ground level.
  const x = origin.x + dir.x * 60;
  const z = origin.z + dir.z * 60;
  const ground = collision.groundHeightAt(x, z, origin.y + 50, 200);
  return vec3(x, Number.isFinite(ground) ? ground : origin.y, z);
}

const _traceHit = {
  hit: false,
  distance: 0,
  point: vec3(),
  normal: vec3(0, 1, 0),
  surface: 0 as never,
  entity: 0 as EntityId,
  brushIndex: -1,
  thickness: 0,
  layer: CollisionLayer.None,
};

function spawnAirVehicle(
  rt: KillstreakRuntime,
  def: KillstreakDef,
  player: PlayerState,
  world: WorldState,
): KillstreakEntityState {
  const kind = def.vehicle ?? KillstreakVehicleKind.Chopper;
  // Enter from the caller's side of the map, high up.
  const entry = vec3(player.position.x * 0.4, 42, player.position.z * 1.4);
  return {
    id: rt.nextId++,
    kind,
    owner: player.id,
    team: player.team,
    position: entry,
    velocity: vec3(),
    yaw: player.yaw,
    pitch: 0,
    health: def.health ?? 900,
    timeRemaining: def.duration,
    controlled: def.kind === 'controlled',
    pathIndex: 0,
  };
  void world;
}

// ---------------------------------------------------------------------------
// Per-tick
// ---------------------------------------------------------------------------

export interface KillstreakTickResult {
  events: SimEvent[];
  /** Explosions to resolve, in world space. */
  explosions: Array<{ position: Vec3; radius: number; damage: number; owner: PlayerId }>;
  /** Direct damage from vehicle weapons. */
  hits: Array<{ victim: PlayerId; attacker: PlayerId; damage: number; position: Vec3 }>;
}

const _tick: KillstreakTickResult = { events: [], explosions: [], hits: [] };
const _toTarget = vec3();

export function stepKillstreaks(
  world: WorldState,
  collision: CollisionWorld,
  rt: KillstreakRuntime,
  dt: number,
  rng: Rng,
): KillstreakTickResult {
  _tick.events = [];
  _tick.explosions = [];
  _tick.hits = [];

  // --- team effects --------------------------------------------------------
  for (const effects of rt.effects.values()) {
    effects.uav = Math.max(0, effects.uav - dt);
    effects.advancedUav = Math.max(0, effects.advancedUav - dt);
    effects.counterUav = Math.max(0, effects.counterUav - dt);
    effects.emp = Math.max(0, effects.emp - dt);
  }

  // --- pending strikes -----------------------------------------------------
  for (let i = rt.pendingStrikes.length - 1; i >= 0; i--) {
    const strike = rt.pendingStrikes[i]!;
    strike.delay -= dt;
    if (strike.delay > 0) continue;

    // Bombs land in sequence, walking along the run's heading. The walk is what
    // makes an airstrike feel like aircraft passing over rather than a single
    // instant deletion of a circle.
    const spread = strike.kind === 'cluster' ? 9 : 7;
    const offset = (strike.fired - (strike.bombs - 1) / 2) * (spread / Math.max(1, strike.bombs - 1)) * 2;

    const lateral = strike.kind === 'cluster' ? rng.signed(spread) : 0;
    const pos = vec3(
      strike.target.x + strike.heading.x * offset - strike.heading.z * lateral,
      strike.target.y + 0.4,
      strike.target.z + strike.heading.z * offset + strike.heading.x * lateral,
    );

    _tick.explosions.push({
      position: pos,
      radius: strike.radius,
      damage: strike.damage,
      owner: strike.owner,
    });
    _tick.events.push({
      type: SimEventType.Explosion,
      tick: world.tick,
      position: pos,
      radius: strike.radius,
      owner: strike.owner,
      kind: 'killstreak',
    } as SimEvent);

    strike.fired++;
    if (strike.fired >= strike.bombs) {
      rt.pendingStrikes.splice(i, 1);
    } else {
      strike.delay = strike.spacing;
    }
  }

  // --- vehicles ------------------------------------------------------------
  for (const vehicle of Array.from(world.killstreakEntities.values())) {
    vehicle.timeRemaining -= dt;

    if (vehicle.timeRemaining <= 0 || vehicle.health <= 0) {
      world.killstreakEntities.delete(vehicle.id);
      if (vehicle.health <= 0) {
        _tick.events.push({
          type: SimEventType.KillstreakDestroyed,
          tick: world.tick,
          player: vehicle.owner,
          team: vehicle.team,
          position: { ...vehicle.position },
          data: { kind: vehicle.kind },
        } as SimEvent);
        _tick.explosions.push({
          position: vehicle.position,
          radius: 6,
          damage: 100,
          owner: 0,
        });
      }
      continue;
    }

    stepVehicle(world, collision, vehicle, dt, rng);
  }

  return _tick;
}

/**
 * Fly a killstreak vehicle and let it shoot.
 *
 * Uncontrolled aircraft orbit the map centre rather than chasing players: an
 * autonomous gunship that tracks you perfectly is not a reward for the caller,
 * it is a punishment for everyone else with no counterplay except waiting.
 */
function stepVehicle(
  world: WorldState,
  collision: CollisionWorld,
  vehicle: KillstreakEntityState,
  dt: number,
  rng: Rng,
): void {
  const orbitRadius = 55;
  const orbitSpeed = 0.22;

  if (!vehicle.controlled) {
    vehicle.pathIndex += orbitSpeed * dt;
    vehicle.position.x = Math.cos(vehicle.pathIndex) * orbitRadius;
    vehicle.position.z = Math.sin(vehicle.pathIndex) * orbitRadius;
    vehicle.position.y = 38;
    vehicle.yaw = vehicle.pathIndex + Math.PI / 2;
  }

  // Fire on a visible enemy, with a deliberately mediocre hit rate.
  const cooldownKey = vehicle.id;
  const last = vehicleFireTimes.get(cooldownKey) ?? 0;
  const interval = vehicle.kind === KillstreakVehicleKind.Chopper ? 0.16 : 0.09;
  if (world.time - last < interval) return;

  const owner = world.players.get(vehicle.owner);
  const team = owner?.team ?? vehicle.team;

  let best: PlayerState | null = null;
  let bestDist = Infinity;
  for (const player of world.players.values()) {
    if (!player.alive || !isEnemyTeam(team, player.team)) continue;
    const d = v3distance(vehicle.position, player.position);
    if (d > 90 || d >= bestDist) continue;
    // Must actually be able to see them — indoors is safe from air support,
    // which is the whole reason indoor routes exist on these maps.
    v3set(_toTarget, player.position.x, player.position.y + 1.2, player.position.z);
    if (!collision.isVisible(vehicle.position, _toTarget, STRIKE_FILTER)) continue;
    best = player;
    bestDist = d;
  }

  if (!best) return;
  vehicleFireTimes.set(cooldownKey, world.time);

  // Aircraft cannons miss a lot; the threat is being forced into cover, not the
  // damage itself.
  if (!rng.chance(0.35)) {
    _tick.events.push({
      type: SimEventType.Impact,
      tick: world.tick,
      position: { x: best.position.x + rng.signed(2), y: best.position.y, z: best.position.z + rng.signed(2) },
      normal: { x: 0, y: 1, z: 0 },
      surface: 0 as never,
      shooter: vehicle.owner,
      penetrated: false,
    } as SimEvent);
    return;
  }

  _tick.hits.push({
    victim: best.id,
    attacker: vehicle.owner,
    damage: vehicle.kind === KillstreakVehicleKind.Chopper ? 28 : 45,
    position: { ...best.position },
  });
}

/** Last fire time per vehicle. Cleared when a match ends. */
const vehicleFireTimes = new Map<EntityId, number>();

export function resetKillstreakRuntime(rt: KillstreakRuntime): void {
  for (const effects of rt.effects.values()) {
    effects.uav = 0;
    effects.advancedUav = 0;
    effects.counterUav = 0;
    effects.emp = 0;
  }
  rt.pendingStrikes.length = 0;
  vehicleFireTimes.clear();
}

// ---------------------------------------------------------------------------
// Queries
// ---------------------------------------------------------------------------

/**
 * Is an enemy visible on a team's minimap right now?
 *
 * Split out because three different systems need the same answer and they must
 * never disagree — the minimap showing someone the AI cannot see, or vice versa,
 * is the kind of inconsistency players notice immediately and cannot explain.
 */
export function isRevealedOnRadar(
  rt: KillstreakRuntime,
  viewerTeam: Team,
  target: PlayerState,
  suppressedRecently: boolean,
): boolean {
  if (!isEnemyTeam(viewerTeam, target.team)) return true;
  if (suppressedRecently) return true;
  return hasRadar(rt, viewerTeam);
}

/** Seconds of radar left, for the HUD's UAV sweep. */
export function radarTimeRemaining(rt: KillstreakRuntime, team: Team): number {
  if (!hasRadar(rt, team)) return 0;
  const e = teamEffects(rt, team);
  return Math.max(e.uav, e.advancedUav);
}

File: dotnet/src/OperationVanguard.Core/Simulation/KillstreakRuntime.cs
namespace OperationVanguard.Core;

/// <summary>Team-wide killstreak effects which have no world entity.</summary>
public sealed class TeamEffects
{
    public double Uav { get; set; }

    public double AdvancedUav { get; set; }

    public double CounterUav { get; set; }

    public double Emp { get; set; }
}

/// <summary>A scripted strike in flight.</summary>
public sealed class PendingStrike
{
    /// <summary>One of "airstrike", "cluster", or "cruise".</summary>
    public string Kind { get; set; } = string.Empty;

    public int Owner { get; set; }

    public Team Team { get; set; }

    public Vec3 Target { get; set; } = new();

    public double Delay { get; set; }

    public double Damage { get; set; }

    public double Radius { get; set; }

    public int Bombs { get; set; }

    public double Spacing { get; set; }

    public int Fired { get; set; }

    public Vec3 Heading { get; set; } = new();
}

/// <summary>State retained by the killstreak runtime across simulation ticks.</summary>
public sealed class KillstreakRuntime
{
    public Dictionary<Team, TeamEffects> Effects { get; set; } = [];

    public List<PendingStrike> PendingStrikes { get; set; } = [];

    public int NextId { get; set; } = 50_000;
}

public sealed class KillstreakResult
{
    public List<SimEvent> Events { get; set; } = [];

    public List<KillstreakEntityState> Spawned { get; set; } = [];

    public bool Used { get; set; }

    public bool EndsMatch { get; set; }
}

public sealed class KillstreakTickResult
{
    public List<SimEvent> Events { get; set; } = [];

    public List<ExplosionRequest> Explosions { get; set; } = [];

    public List<DirectHitRequest> Hits { get; set; } = [];
}

/// <summary>Killstreak activation, lifecycle, and radar queries.</summary>
public static class KillstreakRuntimeSystem
{
    private const string Airstrike = "airstrike";
    private const string Cluster = "cluster";
    private const string Cruise = "cruise";

    private static readonly QueryFilter StrikeFilter =
        new(CollisionLayer.World | CollisionLayer.Breakable);
    private static readonly Vec3 Aim = new();
    private static readonly Vec3 Eye = new();
    private static readonly Vec3 ToTarget = new();
    private static readonly RaycastHit TraceHit = new();

    private static readonly KillstreakResult Result = new();
    private static readonly KillstreakTickResult Tick = new();
    private static readonly Dictionary<int, double> VehicleFireTimes = [];

    public static KillstreakRuntime CreateKillstreakRuntime()
    {
        var runtime = new KillstreakRuntime();
        runtime.Effects.Add(Team.Allies, EmptyEffects());
        runtime.Effects.Add(Team.Axis, EmptyEffects());
        runtime.Effects.Add(Team.None, EmptyEffects());
        return runtime;
    }

    public static TeamEffects TeamEffects(KillstreakRuntime runtime, Team team)
    {
        if (!runtime.Effects.TryGetValue(team, out var effects))
        {
            effects = EmptyEffects();
            runtime.Effects[team] = effects;
        }

        return effects;
    }

    public static bool HasRadar(KillstreakRuntime runtime, Team team)
    {
        var own = TeamEffects(runtime, team);
        if (own.Uav <= 0d && own.AdvancedUav <= 0d)
        {
            return false;
        }

        foreach (var pair in runtime.Effects)
        {
            if (pair.Key == team || !SimulationTypes.IsEnemyTeam(team, pair.Key))
            {
                continue;
            }

            if (pair.Value.CounterUav > 0d)
            {
                return false;
            }
        }

        return true;
    }

    public static KillstreakResult CallKillstreak(
        WorldState world,
        ICollisionWorld collision,
        KillstreakRuntime runtime,
        PlayerState player,
        string streakId,
        Rng rng)
    {
        Result.Events = [];
        Result.Spawned = [];
        Result.Used = false;
        Result.EndsMatch = false;

        if (!KillstreakData.Killstreaks.TryGetValue(streakId, out var definition))
        {
            return Result;
        }

        var inventoryIndex = player.KillstreakInventory.IndexOf(streakId);
        if (inventoryIndex < 0)
        {
            return Result;
        }

        MathEx.Set(Eye, player.Position.X, player.Position.Y + 1.6d, player.Position.Z);
        MathEx.AnglesToForward(Aim, player.Yaw, player.Pitch);
        var target = TraceToGround(collision, Eye, Aim);

        switch (streakId)
        {
            case "uav":
                TeamEffects(runtime, player.Team).Uav = definition.Duration;
                break;

            case "advanced_uav":
                TeamEffects(runtime, player.Team).AdvancedUav = definition.Duration;
                TeamEffects(runtime, player.Team).Uav = definition.Duration;
                break;

            case "counter_uav":
                TeamEffects(runtime, player.Team).CounterUav = definition.Duration;
                break;

            case "emp_burst":
                foreach (var pair in runtime.Effects)
                {
                    if (SimulationTypes.IsEnemyTeam(player.Team, pair.Key))
                    {
                        pair.Value.Emp = definition.Duration;
                    }
                }
                break;

            case "precision_airstrike":
                runtime.PendingStrikes.Add(new PendingStrike
                {
                    Kind = Airstrike,
                    Owner = player.Id,
                    Team = player.Team,
                    Target = target,
                    Delay = 4d,
                    Damage = definition.Damage ?? 180d,
                    Radius = definition.Radius ?? 7d,
                    Bombs = 6,
                    Spacing = 0.18d,
                    Fired = 0,
                    Heading = StrikeHeading(player),
                });
                break;

            case "cluster_strike":
                runtime.PendingStrikes.Add(new PendingStrike
                {
                    Kind = Cluster,
                    Owner = player.Id,
                    Team = player.Team,
                    Target = target,
                    Delay = 3d,
                    Damage = definition.Damage ?? 130d,
                    Radius = definition.Radius ?? 5d,
                    Bombs = 9,
                    Spacing = 0.12d,
                    Fired = 0,
                    Heading = StrikeHeading(player),
                });
                break;

            case "cruise_missile":
                runtime.PendingStrikes.Add(new PendingStrike
                {
                    Kind = Cruise,
                    Owner = player.Id,
                    Team = player.Team,
                    Target = target,
                    Delay = 2.5d,
                    Damage = definition.Damage ?? 220d,
                    Radius = definition.Radius ?? 9d,
                    Bombs = 1,
                    Spacing = 0d,
                    Fired = 0,
                    Heading = StrikeHeading(player),
                });
                break;

            case "sentry_gun":
            case "care_package":
                break;

            case "attack_chopper":
            case "chopper_gunner":
            case "vtol_jet":
            case "gunship":
                Result.Spawned.Add(SpawnAirVehicle(runtime, definition, player, world));
                break;

            case "juggernaut":
                player.MaxHealth = 400d;
                player.Health = 400d;
                break;

            case "tactical_nuke":
                Result.EndsMatch = true;
                break;
        }

        player.KillstreakInventory.RemoveAt(inventoryIndex);
        Result.Used = true;
        Result.Events.Add(new GenericSimEvent(SimEventType.KillstreakCalled)
        {
            Tick = world.Tick,
            Player = player.Id,
            Team = player.Team,
            Data = new Dictionary<string, object?>
            {
                ["killstreakId"] = streakId,
                ["target"] = new Vec3(target.X, target.Y, target.Z),
            },
        });
        Result.Events.Add(new AnnounceEvent
        {
            Tick = world.Tick,
            Team = player.Team,
            Line = definition.FriendlyAnnounce,
        });

        if (!string.IsNullOrEmpty(definition.EnemyAnnounce))
        {
            Result.Events.Add(new AnnounceEvent
            {
                Tick = world.Tick,
                Team = Opposing(player.Team),
                Line = definition.EnemyAnnounce,
            });
        }

        _ = rng;
        return Result;
    }

    public static KillstreakTickResult StepKillstreaks(
        WorldState world,
        ICollisionWorld collision,
        KillstreakRuntime runtime,
        double deltaTime,
        Rng rng)
    {
        Tick.Events = [];
        Tick.Explosions = [];
        Tick.Hits = [];

        foreach (var effects in runtime.Effects.Values)
        {
            effects.Uav = Math.Max(0d, effects.Uav - deltaTime);
            effects.AdvancedUav = Math.Max(0d, effects.AdvancedUav - deltaTime);
            effects.CounterUav = Math.Max(0d, effects.CounterUav - deltaTime);
            effects.Emp = Math.Max(0d, effects.Emp - deltaTime);
        }

        for (var index = runtime.PendingStrikes.Count - 1; index >= 0; index--)
        {
            var strike = runtime.PendingStrikes[index];
            strike.Delay -= deltaTime;
            if (strike.Delay > 0d)
            {
                continue;
            }

            var spread = strike.Kind == Cluster ? 9d : 7d;
            var offset =
                (strike.Fired - (strike.Bombs - 1d) / 2d) *
                (spread / Math.Max(1, strike.Bombs - 1)) *
                2d;
            var lateral = strike.Kind == Cluster ? rng.Signed(spread) : 0d;
            var position = new Vec3(
                strike.Target.X + strike.Heading.X * offset - strike.Heading.Z * lateral,
                strike.Target.Y + 0.4d,
                strike.Target.Z + strike.Heading.Z * offset + strike.Heading.X * lateral);

            Tick.Explosions.Add(new ExplosionRequest
            {
                Position = position,
                Radius = strike.Radius,
                Damage = strike.Damage,
                Owner = strike.Owner,
            });
            Tick.Events.Add(new ExplosionEvent
            {
                Tick = world.Tick,
                Position = position,
                Radius = strike.Radius,
                Owner = strike.Owner,
                Kind = "killstreak",
            });

            strike.Fired++;
            if (strike.Fired >= strike.Bombs)
            {
                runtime.PendingStrikes.RemoveAt(index);
            }
            else
            {
                strike.Delay = strike.Spacing;
            }
        }

        foreach (var vehicle in world.KillstreakEntities.Values.ToArray())
        {
            vehicle.TimeRemaining -= deltaTime;
            if (vehicle.TimeRemaining <= 0d || vehicle.Health <= 0d)
            {
                world.KillstreakEntities.Remove(vehicle.Id);
                if (vehicle.Health <= 0d)
                {
                    Tick.Events.Add(new GenericSimEvent(SimEventType.KillstreakDestroyed)
                    {
                        Tick = world.Tick,
                        Player = vehicle.Owner,
                        Team = vehicle.Team,
                        Position = new Vec3(vehicle.Position.X, vehicle.Position.Y, vehicle.Position.Z),
                        Data = new Dictionary<string, object?> { ["kind"] = vehicle.Kind },
                    });
                    Tick.Explosions.Add(new ExplosionRequest
                    {
                        Position = vehicle.Position,
                        Radius = 6d,
                        Damage = 100d,
                        Owner = 0,
                    });
                }

                continue;
            }

            StepVehicle(world, collision, vehicle, deltaTime, rng);
        }

        return Tick;
    }

    public static void ResetKillstreakRuntime(KillstreakRuntime runtime)
    {
        foreach (var effects in runtime.Effects.Values)
        {
            effects.Uav = 0d;
            effects.AdvancedUav = 0d;
            effects.CounterUav = 0d;
            effects.Emp = 0d;
        }

        runtime.PendingStrikes.Clear();
        VehicleFireTimes.Clear();
    }

    public static bool IsRevealedOnRadar(
        KillstreakRuntime runtime,
        Team viewerTeam,
        PlayerState target,
        bool suppressedRecently)
    {
        if (!SimulationTypes.IsEnemyTeam(viewerTeam, target.Team))
        {
            return true;
        }

        if (suppressedRecently)
        {
            return true;
        }

        return HasRadar(runtime, viewerTeam);
    }

    public static double RadarTimeRemaining(KillstreakRuntime runtime, Team team)
    {
        if (!HasRadar(runtime, team))
        {
            return 0d;
        }

        var effects = TeamEffects(runtime, team);
        return Math.Max(effects.Uav, effects.AdvancedUav);
    }

    private static TeamEffects EmptyEffects() => new();

    private static Team Opposing(Team team) =>
        team == Team.Allies ? Team.Axis : Team.Allies;

    private static Vec3 StrikeHeading(PlayerState player)
    {
        var heading = new Vec3();
        MathEx.AnglesToForward(heading, player.Yaw, 0d);
        return heading;
    }

    private static Vec3 TraceToGround(ICollisionWorld collision, Vec3 origin, Vec3 direction)
    {
        var hit = collision.Raycast(origin, direction, 300d, StrikeFilter, TraceHit);
        if (hit.Hit)
        {
            return new Vec3(hit.Point.X, hit.Point.Y, hit.Point.Z);
        }

        var x = origin.X + direction.X * 60d;
        var z = origin.Z + direction.Z * 60d;
        var ground = collision.GroundHeightAt(x, z, origin.Y + 50d, 200d);
        return new Vec3(x, double.IsFinite(ground) ? ground : origin.Y, z);
    }

    private static KillstreakEntityState SpawnAirVehicle(
        KillstreakRuntime runtime,
        KillstreakDef definition,
        PlayerState player,
        WorldState world)
    {
        var kind = definition.Vehicle ?? KillstreakVehicleKind.Chopper;
        var entry = new Vec3(player.Position.X * 0.4d, 42d, player.Position.Z * 1.4d);
        var entity = new KillstreakEntityState
        {
            Id = runtime.NextId++,
            Kind = kind,
            Owner = player.Id,
            Team = player.Team,
            Position = entry,
            Velocity = new Vec3(),
            Yaw = player.Yaw,
            Pitch = 0d,
            Health = definition.Health ?? 900d,
            TimeRemaining = definition.Duration,
            Controlled = definition.Kind == KillstreakKind.Controlled,
            PathIndex = 0d,
        };
        _ = world;
        return entity;
    }

    private static void StepVehicle(
        WorldState world,
        ICollisionWorld collision,
        KillstreakEntityState vehicle,
        double deltaTime,
        Rng rng)
    {
        const double orbitRadius = 55d;
        const double orbitSpeed = 0.22d;
        if (!vehicle.Controlled)
        {
            vehicle.PathIndex += orbitSpeed * deltaTime;
            vehicle.Position.X = Math.Cos(vehicle.PathIndex) * orbitRadius;
            vehicle.Position.Z = Math.Sin(vehicle.PathIndex) * orbitRadius;
            vehicle.Position.Y = 38d;
            vehicle.Yaw = vehicle.PathIndex + Math.PI / 2d;
        }

        var last = VehicleFireTimes.TryGetValue(vehicle.Id, out var value) ? value : 0d;
        var interval = vehicle.Kind == KillstreakVehicleKind.Chopper ? 0.16d : 0.09d;
        if (world.Time - last < interval)
        {
            return;
        }

        var team = world.Players.TryGetValue(vehicle.Owner, out var owner)
            ? owner.Team
            : vehicle.Team;
        PlayerState? best = null;
        var bestDistance = double.PositiveInfinity;
        foreach (var player in world.Players.Values)
        {
            if (!player.Alive || !SimulationTypes.IsEnemyTeam(team, player.Team))
            {
                continue;
            }

            var distance = MathEx.Distance(vehicle.Position, player.Position);
            if (distance > 90d || distance >= bestDistance)
            {
                continue;
            }

            MathEx.Set(ToTarget, player.Position.X, player.Position.Y + 1.2d, player.Position.Z);
            if (!collision.IsVisible(vehicle.Position, ToTarget, StrikeFilter))
            {
                continue;
            }

            best = player;
            bestDistance = distance;
        }

        if (best is null)
        {
            return;
        }

        VehicleFireTimes[vehicle.Id] = world.Time;
        if (!rng.Chance(0.35d))
        {
            Tick.Events.Add(new ImpactEvent
            {
                Tick = world.Tick,
                Position = new Vec3(
                    best.Position.X + rng.Signed(2d),
                    best.Position.Y,
                    best.Position.Z + rng.Signed(2d)),
                Normal = new Vec3(0d, 1d, 0d),
                Surface = default,
                Shooter = vehicle.Owner,
                Penetrated = false,
            });
            return;
        }

        Tick.Hits.Add(new DirectHitRequest
        {
            Victim = best.Id,
            Attacker = vehicle.Owner,
            Damage = vehicle.Kind == KillstreakVehicleKind.Chopper ? 28d : 45d,
            Position = new Vec3(best.Position.X, best.Position.Y, best.Position.Z),
        });
    }
}
